common/transport/amqp/src/sender_link.ts

```diff
@@ -0,0 +1,207 @@
+                            .then((state) => {
+                              _processPendingMessageCallback(null, new results.MessageEnqueued(state));
+                              return null;
+                            })
+                            .catch((err) => {
+                              _processPendingMessageCallback(err);
+                            });
+          }
+        },
+        detaching: {
+          _onEnter: (err) => {
+            if (this._linkObject) {
+              /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_009: [** The `detach` method shall detach the link created by the `amqp10.AmqpClient` underlying object.]*/
+              this._linkObject.forceDetach();
+              this._linkObject = null;
+            }
+            /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_017: [** The `detach` method shall return the state machine to the `detached` state.]*/
+            this._fsm.transition('detached', null, err);
+          },
+          '*': () => this._fsm.deferUntilTransition('detached')
+        }
+      }
+    });
+  }
+
+  detach(): void {
+    this._fsm.handle('detach');
+  }
+
+  attach(callback: (err?: Error) => void): void {
+    this._fsm.handle('attach', callback);
+  }
+
+  send(message: AmqpMessage, callback: (err?: Error, result?: results.MessageEnqueued) => void): void {
+    this._fsm.handle('send', message, callback);
+  }
+
+  private _attachLink(callback: (err?: Error) => void): void {
+    /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_008: [If the `amqp10.AmqpClient` fails to create the link the `callback` function shall be called with this error object.]*/
+    let connectionError = null;
+    let clientErrorHandler = (err) => {
+      connectionError = err;
+    };
+
+    /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_005: [If the `amqp10.AmqpClient` emits an `errorReceived` event during the time the link is attached, the `callback` function shall be called with this error.]*/
+    this._amqp10Client.on('client:errorReceived', clientErrorHandler);
+
+    /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_004: [The `attach` method shall use the stored instance of the `amqp10.AmqpClient` object to attach a new link object with the `linkAddress` and `linkOptions` provided when creating the `SenderLink` instance.]*/
+    this._amqp10Client.createSender(this._linkAddress, this._linkOptions)
+      .then((amqp10link) => {
+        if (!connectionError) {
+          debug('Sender object created for endpoint: ' + this._linkAddress);
+          this._linkObject = amqp10link;
+          /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_006: [The `SenderLink` object should subscribe to the `detached` event of the newly created `amqp10` link object.]*/
+          this._linkObject.on('detached', this._detachHandler);
+          /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_007: [The `SenderLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object.]*/
+          this._linkObject.on('errorReceived', this._errorHandler);
```

Comment:
yes.

---
In reply to: [127619882](https://github.com/Azure/azure-iot-sdk-node/pull/55#discussion_r127619882) [](ancestors = 127619882)